scalyr_agent/log_processing.py

```diff
@@ -79,6 +79,44 @@
 
 log = scalyr_logging.getLogger(__name__)
 
+def _parse_cri_log( line ):
+    """
+    Parse a log line that uses the K8S Container Runtime Interface (CRI) format
+    https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-cri-logging.md
+    """
+    # find the end of the timestamp segment
+    index = line.find( ' ' )
+    if index < 0:
+        return None, None, None, None
+
+    #parse the timestamp
+    timestamp = scalyr_util.rfc3339_to_nanoseconds_since_epoch( line[:index] )
+    if timestamp is None:
+        return None, None, None, None
+
+    # skip over the ' ' delimiter and parse the stream type
+    line = line[index+1:]
+    index = line.find( ' ' )
+    if index < 0:
+        return None, None, None, None
+
+    stream = line[:index]
+    if stream != 'stdout' and stream != 'stderr':
+        return None, None, None, None
+
+    # skip over the ' ' delimiter and parse the tags
+    line = line[index+1:]
+    index = line.find( ' ' )
+    if index < 0:
+        return None, None, None, None
+
+    tags = line[:index]
+
+    # skip over the ' ' delimiter and get the log line
+    line = line[index+1:]
+
+    return timestamp, stream, tags, line
+
```

Comment:
For context, the current code is almost a line for line conversion of the [Go code used by k8s](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/logs/logs.go#L125-L169) to parse CRI logs.

There is no issue with multiple spaces, because the format currently uses single space as the delimiter.

Given that there isn't really a clear specification for this log format outside of the code, I think it's clearer to have variables explicitly named so people reading the code can easily read what each slice (and return value) represents.

`return s[0:locs[1]], s[locs[1]+1:locs[2]], s[locs[2]+1:locs[3]], s[locs[3]+1:]` is significantly less readable than `return timestamp, stream, tags, line`

Having them all bunched up like that also makes it easier for errors to slip in, for example running the suggested replacement there appears to be an out of bounds error because after getting rid of the initial -1, the array only contains 3 elements, zero indexed, and they are being accessed as if they are 1-indexed.  An easy issue to fix, but it does illustrate the point about how complexity makes it easy for bugs to creep in.

Finally, I'm not sure the concern about performance is warranted.  This code *is* on a hot path, but that path has been profiled heavily and running fine previously doing full json-parsing on log lines (which by default are json objects).  Previous profiling revealed that turning off json parsing and just uploading the raw line didn't have much impact on overall performance (for reference, the biggest hit to performance on hot paths is debug statements that are never printed).  

I'd also caution against any performance issues without any profiling to compare versions because it can reveal interesting things.   In the most recent round of performance improvments, the performance hit was on something completely unexepected (the json response from the docker/k8s apis didn't specify an encoding so python was doing encoding detection on the returned string (some 2MB+) and that was causing long delays).

Out of curiousity, I did some quick profiling of both the current version of this function and the suggested changes (with the OOB error fixed) and it revealed:

* string slicing has negliable impact on performance here - the biggest hit to performance in that function is the conversion of the timestamp to nanoseconds and the amount of slicing going on is insignificant compared to that.

* when removing the conversion to nanoseconds and just returning the raw string for that field, the original version of the code runs faster than the suggested changes (twice as fast on my machine).

I'll send you the test harness I used separately so you can check the results yourself, but assuming I haven't made an error in how I was timing it (just using timeit), then given that the original code is easier to read and also appears to be faster, I think it's probably ok to leave it as is.

Reply:
Good points.   The locs-array was just an example you can of course assign explicit variables to the location indexes, as well as the slices, and return a 4-tuple of nice variable names.  

Readability matters aside, I fully agree about avoiding premature performance optimization.   However, my concern is not performance (speed), rather memory leaks.  IIUC we have had past problems where the cause is still unknown and the fix was to explicitly force garbage collection. 

With this in mind, I feel further discussion on this point is warranted:

- Go creates slice-by-reference.  It [does not copy data](https://blog.golang.org/go-slices-usage-and-internals).
- In contrast, python makes [a completely new copy](https://stackoverflow.com/questions/5722006/does-python-do-slice-by-reference-on-strings).

Imagine a situation where these logs are heavy and each line is long -- say 1000 chars. The go example will chop off a few chars for the metadata columns.  And each time it does, it creates entirely new strings, say, 990, 980, 970 chars long.  Multiply that by a large number of lines and you get the picture.  

In contrast, the approach I propose is simply slice exactly what you need without creating these intermediate long strings:-
3 short slices for the metadata
A single 970 char slice for the log line.

In summary,  I feel there is really no downside and this isn't an optimization per-se, rather avoiding creating unnecessary objects given how Python string slices work.

Reply:
Are we still going to reduce the number of slices as discussed, or were you planning on addressing this in a future PR?

Reply:
I did some memory testing with the `guppy` lib and heap usage appears to be identical for both versions.  I'll send you the test harness so you can confirm, but if it looks ok then considering the original is slightly faster, I think we can leave as is for now.

Reply:
sgtm.